lighthouse-core/gather/computed/first-interactive.js

```diff
@@ -0,0 +1,206 @@
+    const tInSeconds = t / 1000;
+    const exponentiationComponent = Math.exp(-.045 * tInSeconds);
+    return (4 * exponentiationComponent + 1) * 1000;
+  }
+
+  /**
+   * @param {!Array<{start: number, end: number}>} tasks
+   * @param {number} startIndex
+   * @param {number} windowEnd
+   * @return {!Array<{start: number, end: number, duration: number}>}
+   */
+  static getTaskClustersInWindow(tasks, startIndex, windowEnd) {
+    const clusters = [];
+
+    let previousTaskEndTime = -Infinity;
+    let currentCluster = null;
+
+    // Examine all tasks that could possibly be part of a cluster starting before windowEnd.
+    // Consider the case where window end is 15s, there's a 100ms task from 14.9-15s and a 500ms
+    // task from 15.5-16s, we need that later task to be clustered with the first so we can properly
+    // identify that main thread isn't quiet.
+    const clusteringWindowEnd = windowEnd + MIN_TASK_CLUSTER_PADDING;
+    const couldTaskBelongToCluster = task => task && task.start < clusteringWindowEnd;
+    for (let i = startIndex; couldTaskBelongToCluster(tasks[i]); i++) {
+      const task = tasks[i];
+
+      // if enough time has elapsed, we'll create a new cluster
+      if (task.start - previousTaskEndTime > MIN_TASK_CLUSTER_PADDING) {
+        if (currentCluster) {
+          clusters.push(currentCluster);
+        }
+
+        currentCluster = [];
+      }
+
+      currentCluster.push(task);
+      previousTaskEndTime = task.end;
+    }
+
+    if (currentCluster) {
+      clusters.push(currentCluster);
+    }
+
+    return clusters
+      // add some useful information about the cluster
+      .map(tasks => {
+        const start = tasks[0].start;
+        const end = tasks[tasks.length - 1].end;
+        const duration = end - start;
+        return {start, end, duration, tasks};
+      })
+      // filter out clusters that started after the window because of our clusteringWindowEnd
+      .filter(cluster => cluster.start < windowEnd);
+  }
+
+  /**
+   * @param {number} FMP
+   * @param {number} traceEnd
+   * @param {!Array<{start: number, end: number>}} longTasks
+   * @return {number}
+   */
+  static findQuietWindow(FMP, traceEnd, longTasks) {
+    if (longTasks.length === 0 ||
+        longTasks[0].start > FMP + MAX_QUIET_WINDOW_SIZE) {
+      return FMP;
+    }
+
+    // FirstInteractive must start at the end of a long task, consider each long task and
+    // examine the window that follows it.
+    for (let i = 0; i < longTasks.length; i++) {
+      const event = longTasks[i];
+      const windowStart = event.end;
+      const windowSize = FirstInteractive.getRequiredWindowSizeInMs(windowStart - FMP);
+      const windowEnd = windowStart + windowSize;
+
+      // Check that we have a long enough trace
+      if (windowEnd > traceEnd) {
+        throw new Error(TRACE_BUSY_MSG);
+      }
+
+      const isTooCloseToFMP = cluster => cluster.start < FMP + MIN_TASK_CLUSTER_FMP_DISTANCE;
```

Comment:
moved outside the loop, they're separate for readability and clarity of each condition that must be true